### ISSUE TITLE
Treatment of hyphenated words

### ISSUE THREAD

**Issue by rattle99** [huggingface/tokenizers#1507]
It seems huggingface tokenizers, treats hyphenated words as separate, including the hyphen with reference to the `word_ids()` function.

For example, in the sentence

> 'To win the money , SpaceShipOne had to blast off into space twice in a two-week period and fly at least 100 kilometers above Earth .'

Using 
```py
tokenizer = AutoTokenizer.from_pretrained('distilbert-base-uncased')

text = "To win the money , SpaceShipOne had to blast off into space twice in a two-week period and fly at least 100 kilometers above Earth ."

text_tokenized = tokenizer(text , padding = 'longest' , truncation = True, return_tensors = "pt", is_split_into_words=False )
print(text_tokenized.word_ids())

```
returns
`[None, 0, 1, 2, 3, 4, 5, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, None]`

but changing `two-week` to `twoweek` changes it to
`[None, 0, 1, 2, 3, 4, 5, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, None]`

It is possible that this behavior is seen for symbols other than hyphen as well which is worth keeping in mind while doing NER tasks. Or perhaps this is by design?

**Comment by ArthurZucker** [comment]
Hey I think this is heavily related to the design of Bert

**Comment by rattle99** [comment]
> Hey I think this is heavily related to the design of Bert

Hi @ArthurZucker , true but the word_ids() function is agnostic to the model used I think. Perhaps using token spans returned from any tokenizer is a better approach.

**Comment by ArthurZucker** [comment]
word_ids() is dependent on the pre_tokenizer, normalizer etc etc

**Comment by rattle99** [comment]
> word_ids() is dependent on the pre_tokenizer, normalizer etc etc

I see, that explains why. Thank you!

**Comment by ArthurZucker** [comment]
But it is my job to make sure it works as inteded, I know gemma ones are broken, I'll try to spend time on fixing this!

**Comment by rattle99** [comment]
Thank you!
Here is a little code snippet that I worked on for my specific use case when I came across this problem. In case it's helpful.

```
tokenizer = AutoTokenizer.from_pretrained(CONFIG['model_id'])

def get_bio_tags(text, spans):
    tokenized_inputs = tokenizer(text, 
                                 return_offsets_mapping=True, 
                                 add_special_tokens=False, 
                                 return_attention_mask=False,
                                 max_length=510,
                                 truncation=True)
    tokens = tokenizer.convert_ids_to_tokens(tokenized_inputs['input_ids'])
    token_spans = tokenized_inputs['offset_mapping']

    # Initialize BIO tags with "O" (outside) for each token
    bio_tags = ["O"] * len(tokens)

    # Adjust spans to account for special tokens
    for start, end, label in spans:
        is_label_started = False
        if label == 'O':
            continue
        for i, (token_start, token_end) in enumerate(token_spans):
            if token_start == token_end:  # Skip special tokens with no offset
                continue
            if token_start >= start and token_end <= end:
                if not is_label_started:
                    bio_tags[i] = f"B-{label}"
                    is_label_started = True
                else:
                    bio_tags[i] = f"I-{label}"

    return tokens, bio_tags

# Example usage
text = "Dear Omer, as per our records, your license 78B5R2MVFAHJ48500 is still registered in our records for access to the educational tools. Please feedback on it's operability."
spans = [[0, 5, "O"], [5, 9, "FIRSTNAME"], [9, 44, "O"], [44, 61, "VEHICLEVIN"], [61, 170, "O"]]

tokens, bio_tags = get_bio_tags(text, labels)

for token, tag in zip(tokens, bio_tags):
    print(f"{token}\t-->\t{tag}")

```

It was for a NER task.
